source/socket.c

```diff
@@ -0,0 +1,259 @@
+    AWS_PRECONDITION(socket->vtable && socket->vtable->socket_set_options_fn);
+    return socket->vtable->socket_set_options_fn(socket, options);
+}
+
+int aws_socket_assign_to_event_loop(struct aws_socket *socket, struct aws_event_loop *event_loop) {
+    AWS_PRECONDITION(socket->vtable && socket->vtable->socket_assign_to_event_loop_fn);
+    return socket->vtable->socket_assign_to_event_loop_fn(socket, event_loop);
+}
+
+struct aws_event_loop *aws_socket_get_event_loop(struct aws_socket *socket) {
+    return socket->event_loop;
+}
+
+int aws_socket_subscribe_to_readable_events(
+    struct aws_socket *socket,
+    aws_socket_on_readable_fn *on_readable,
+    void *user_data) {
+    AWS_PRECONDITION(socket->vtable && socket->vtable->socket_subscribe_to_readable_events_fn);
+    return socket->vtable->socket_subscribe_to_readable_events_fn(socket, on_readable, user_data);
+}
+
+int aws_socket_read(struct aws_socket *socket, struct aws_byte_buf *buffer, size_t *amount_read) {
+    AWS_PRECONDITION(socket->vtable && socket->vtable->socket_read_fn);
+    return socket->vtable->socket_read_fn(socket, buffer, amount_read);
+}
+
+int aws_socket_write(
+    struct aws_socket *socket,
+    const struct aws_byte_cursor *cursor,
+    aws_socket_on_write_completed_fn *written_fn,
+    void *user_data) {
+
+    AWS_PRECONDITION(socket->vtable && socket->vtable->socket_write_fn);
+    return socket->vtable->socket_write_fn(socket, cursor, written_fn, user_data);
+}
+
+int aws_socket_get_error(struct aws_socket *socket) {
+    AWS_PRECONDITION(socket->vtable && socket->vtable->socket_get_error_fn);
+    return socket->vtable->socket_get_error_fn(socket);
+}
+
+bool aws_socket_is_open(struct aws_socket *socket) {
+    AWS_PRECONDITION(socket->vtable && socket->vtable->socket_is_open_fn);
+    return socket->vtable->socket_is_open_fn(socket);
+}
+
+/**
+ * Return the default socket implementation type. If the return value is `AWS_SOCKET_IMPL_PLATFORM_DEFAULT`, the
+ * function failed to retrieve the default type value.
+ */
+static enum aws_socket_impl_type aws_socket_get_default_impl_type(void) {
+    enum aws_socket_impl_type type = AWS_SOCKET_IMPL_PLATFORM_DEFAULT;
+// override default socket
+#ifdef AWS_USE_APPLE_NETWORK_FRAMEWORK
+    type = AWS_SOCKET_IMPL_APPLE_NETWORK_FRAMEWORK;
+#endif // AWS_USE_APPLE_NETWORK_FRAMEWORK
+    if (type != AWS_SOCKET_IMPL_PLATFORM_DEFAULT) {
+        return type;
+    }
+/**
+ * Ideally we should use the platform definition (e.x.: AWS_OS_APPLE) here, however the platform
+ * definition was declared in aws-c-common. We probably do not want to introduce extra dependency here.
+ */
+#if defined(AWS_ENABLE_KQUEUE) || defined(AWS_ENABLE_EPOLL)
+    return AWS_SOCKET_IMPL_POSIX;
+#elif AWS_ENABLE_DISPATCH_QUEUE
+    return AWS_SOCKET_IMPL_APPLE_NETWORK_FRAMEWORK;
+#elif AWS_ENABLE_IO_COMPLETION_PORTS
+    return AWS_SOCKET_IMPL_WINSOCK;
+#else
+    return AWS_SOCKET_IMPL_PLATFORM_DEFAULT;
```

Comment:
Is this reachable?  Don't we cmake error if no event loop types are enabled?  Maybe add a fatal assert